operator/internal/handlers/lokistack_create_or_update.go

```diff
@@ -279,6 +280,42 @@ func CreateOrUpdateLokiStack(
 		}
 	}
 
+	if stack.Spec.Replication != nil && len(stack.Spec.Replication.Zones) > 0 {
+		podList := &corev1.PodList{}
+		if err = k.List(ctx, podList, &client.ListOptions{
+			Namespace: stack.Namespace,
+			LabelSelector: labels.SelectorFromSet(labels.Set{
+				"app.kubernetes.io/instance": stack.Name,
+			}),
+		}); err != nil {
+			ll.Error(err, "Error getting pod resources")
+			return nil
+		}
+
+		for _, pod := range podList.Items {
+			assignedNode := &corev1.Node{}
+			key := client.ObjectKey{Name: pod.Spec.NodeName}
+			if err = k.Get(ctx, key, assignedNode); err != nil {
+				ll.Error(err, "Error getting node resources in event handler")
+				return nil
+			}
+
+			availabilityZone, er := GetAvailabilityZone([]string{corev1.LabelTopologyZone}, assignedNode.Labels)
+			if er != nil {
+				ll.Error(er, "failed to get pod availability zone from assigned node", "name", pod.Name)
+				return nil
+			}
+
+			if availabilityZone != pod.Annotations[corev1.LabelTopologyZone] {
+				return &status.DegradedError{
+					Message: "Availability zone pod annotation does not match it's node's labels",
+					Reason:  lokiv1.ReasonAvailabilityZoneLabelsMismatch,
+					Requeue: false,
+				}
+			}
+		}
+	}
```

Comment:
Can we make this whole block a function instead of adding it inline here. This handler is already bloated.